Dimmer.cpp

```diff
@@ -1,178 +1,152 @@
  
+ Version 2.0 - Multi-instance objects.
+
  Copyright (c) 2015 Circuitar
  This software is released under the MIT license. See the attached LICENSE file for details.
  */
 
 #include "Arduino.h"
-#include <TimerOne.h>
 #include "Dimmer.h"
 
-//Time table to set triac
-const short triac_time[] ={
-    160, 146, 142, 138, 134, 132, 130, 128, 126, 124,
-    122, 121, 120, 118, 117, 116, 114, 113, 112, 111,
-    110, 109, 108, 107, 106, 105, 104, 103, 102, 101,
-    100,  99,  98,  97,  96,  95,  94,  93,  92,  91,
-    90,  89,  88,  87,  86,  85,  84,  83,  82,  81,
-    80,  79,  78,  77,  76,  75,  74,  73,  72,  71,
-    70,  69,  68,  67,  66,  65,  64,  63,  62,  61,
-    60,  59,  58,  57,  56,  55,  54,  53,  52,  51,
-    50,  48,  47,  46,  45,  44,  42,  41,  40,  38,
-    36,  34,  32,  30,  28,  26,  24,  22,  18,  12, 6
+//Time table to set triacs
+static uint8_t triac_time[] ={
+	160, 146, 142, 138, 134, 132, 130, 128, 126, 124,
+	122, 121, 120, 118, 117, 116, 114, 113, 112, 111,
+	110, 109, 108, 107, 106, 105, 104, 103, 102, 101,
+	100,  99,  98,  97,  96,  95,  94,  93,  92,  91,
+	90,  89,  88,  87,  86,  85,  84,  83,  82,  81,
+	80,  79,  78,  77,  76,  75,  74,  73,  72,  71,
+	70,  69,  68,  67,  66,  65,  64,  63,  62,  61,
+	60,  59,  58,  57,  56,  55,  54,  53,  52,  51,
+	50,  48,  47,  46,  45,  44,  42,  41,  40,  38,
+	36,  34,  32,  30,  28,  26,  24,  22,  18,  12, 6
 };
 
+static DimmerControl* Dimmers[MAX_TRIAC]; // Pointers to all registered dimmer objects
+static uint8_t DimmerCount = 0;           // Number of registered dimmer objects
+static bool started = false;
+
 //ZeroCross Interrupt
 void callZeroCross(){
-    dimmerObject.zeroCross();
-}
-
-//Timer interrupt
-void callTriac(){
-    dimmerObject.triac();
+	for (uint8_t i = 0; i < DimmerCount; i++) {
+		Dimmers[i]->zeroCross();
+	}
 }
 
 //Class Constructor
-Dimmer::Dimmer(){
-    zeroCrosAttached=false;
-    triacAttached=false;
-    triacs=0;
-}
-
-void Dimmer::attachZeroCross(int zc_pin, int inter){
-    zeroCrosAttached=true;
-    pinMode(zc_pin, INPUT);
-    attachInterrupt(inter, callZeroCross, RISING);
-}
-
-void Dimmer::attachTriac(byte pin){
-    triacAttached=true;
-    triacPins[triacs]= pin;
-    triacs++;
-}
-
-int Dimmer::init(){
-    ramp_mode=false;
-    countMode=false;
-    return Dimmer::start();
-}
-
-int Dimmer::initRamp(){
-    ramp_mode=true;
-    countMode=false;
-    return Dimmer::start();
-}
-
-int Dimmer::initCount(byte resolution){
-    ramp_mode=false;
-    countMode=true;
-    countResolution = resolution;
-    return Dimmer::start();
-}
-
-void Dimmer::off(){
-    for(byte i=0; i < triacs; i++){
-        lampSwitch[i]=false;
-    }
-}
+DimmerControl::DimmerControl(uint8_t triacPin, uint8_t value, bool state){
+	this->triacPin = triacPin;
+	this->lampValue = value;
+    this->lampState = state;
 
-void Dimmer::off(byte lamp){
-    lampSwitch[--lamp]=false;
+	//Register dimmer object
+	Dimmers[DimmerCount++] = this;
```

Comment:
I think we should start to check that `DimmerCount` doesn't exceed `MAX_TRIACS`.